ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs

```diff
@@ -682,3 +687,19 @@ getPeersFromCurrentLedger ::
   -> STM m [(PoolStake, NonEmpty DomainAddress)]
 getPeersFromCurrentLedger kernel =
     getPeers . ledgerState <$> ChainDB.getCurrentLedger (getChainDB kernel)
+
+-- | Return the time corresponding to the given slot.
+--
+-- If the slot is too far in the future w.r.t. the current ledger state,
+-- 'HF.PastHorizonException' will be returned.
+getSlotTime ::
```

Comment:
WRONG: we can't use the current `LedgerState` if we don't know anything about the header or block that has that `SlotNo`. In particular, the safe zone that is appropriate for a header should extend from the intersection with our and their chain, not from our tip. Since we don't know the intersection point here, we might be incorrectly returning a `UTCTime` while we should be returning a `PastHorizonException`.

For example, if the `SlotNo` is of a header on a chain that forked off before the transition to Shelley on mainnet and in that chain the transition to Shelley happened ten epochs later, this conversion is completely wrong.

Alternative: introduce a `Timed a` which stores the header and the time corresponding to its slot (computed using the ledger state at the intersection!). It can also include the time at which the header was received by the ChainSyncClient.

Reply:
Who knew time was tricky? :thinking: :grimacing: